Report some read errors back to java code.

hiddenalpha opened the pull request.

hiddenalpha committed:
Report some read errors back to java code.

Relates to https://github.com/java-native/jssc/issues/126

Taken from 19610bcc170c803d9e98030674e377429bfbac03.

tresf commented on src/main/cpp/_nix_based/jssc.cpp:
```diff
@@ -549,22 +549,51 @@ JNIEXPORT jboolean JNICALL Java_jssc_SerialNativeInterface_writeBytes
-    //assert(fd < FD_SETSIZE); // <- Might help when hunting SEGFAULTs.
+    if( fd >= FD_SETSIZE ){
+        jclass exClz = env->FindClass("java/lang/UnsupportedOperationException");
+        if( exClz != NULL ) env->ThrowNew(exClz, "Bad luck. 'select' cannot hanlde large fds.");
```
```suggestion
        if( exClz != NULL ) env->ThrowNew(exClz, "Bad luck. 'select' cannot handle large fds.");
```

tresf commented on src/main/cpp/_nix_based/jssc.cpp:
```diff
@@ -580,17 +609,32 @@ static void awaitReadReady(JNIEnv*, jlong fd){
+                exClz = env->FindClass("java/lang/IllegalArgumentException");
+                if( exClz != NULL ) env->ThrowNew(exClz, "EINVAL poll()");
+                static_assert(EINVAL > 0, "EINVAL > 0");
+                return -EINVAL;
```
Just code consistency with your other switch case.

```suggestion
                return -err;
```

tresf commented on src/main/cpp/_nix_based/jssc.cpp:
```diff
@@ -549,22 +549,51 @@ JNIEXPORT jboolean JNICALL Java_jssc_SerialNativeInterface_writeBytes
+        if( result < 0 ){
+            err = errno;
+            jclass exClz = NULL;
+            switch( err ){
```
The `case` statements should probably be indented to match the styling of the rest of the file.

tresf commented on src/test/java/jssc/SerialNativeInterfaceTest.java:
```diff
@@ -50,4 +58,14 @@ public void reportsWriteErrorsAsIOException() throws Exception {
+    @Test
+    public void throwsIllegalArgumentExceptionIfPortHandleIllegal() throws Exception {
+        try{
+            testTarget.readBytes(999, 42);
```
Should probably be skipped on Windows.

tresf commented on src/test/java/jssc/SerialNativeInterfaceTest.java:
```diff
@@ -50,4 +59,16 @@ public void reportsWriteErrorsAsIOException() throws Exception {
 
+    @Test
+    public void throwsIllegalArgumentExceptionIfPortHandleIllegal() throws Exception {
+        assumeFalse(SerialNativeInterface.getOsType() == SerialNativeInterface.OS_WINDOWS);
```
Now that you've implemented the exception in `windows/jssc.cpp`'s readBytes, should we let this test run?

tresf commented:
For some reason the Intel Mac job keeps getting hung in CI.  I'm going to see if I can determine if it's related to this PR or not before merging.

tresf commented:
The hang definitely appears to occur during `SerialNativeInterfaceTest`.... What's weird is the CI just completely hangs and all of the `mvn` logs vanish once it gets to that point, as if we're somehow conflicting with the pipe... I'll see if I can reproduce locally.

tresf commented:
I can reproduce the hang locally.  Investigating.

tresf commented:
@hiddenalpha although I can reproduce the deadlock, I'm not sure why it's happening...

* On MacOS, when the test `throwsIllegalArgumentExceptionIfPortHandleIllegal` runs, it hangs indefinitely.
* I tried to add some `printf` statements to the C++ code, but could not find a way to display them from Junit.
* I tried to break when `result == 0`, but it doesn't affect the deadlock.
* I tried to provide different `portHandle` values, but they either behave the same or fail the test.

We may decide to just skip this test for MacOS, but before we make that decision, I'd like to know why.  I can offer a Mac environment via SSH if needed.

pietrygamat commented:
The hanging test on MacOS seems related to `poll` vs `select` switch that causes MacOS use different path. From what I could tell, it is not related to the C code changes - the same test on master would get stuck the same way. Defning `#define HAVE_POLL 1` for [__APPLE__](https://github.com/java-native/jssc/blob/0cd6aef0ddf5e8fed6a450906b2f8afa6404086b/src/main/cpp/_nix_based/jssc.cpp#L46) as well as for other OSes results in [a test pass](https://github.com/pietrygamat/jssc/actions/runs/6936752651/job/18869498291), but that gets us back this:
```
    // Seems as poll has some portability issues on OsX (Search for "poll" in
    // "https://cr.yp.to/docs/unixport.html"). So we only make use of poll on
    // all platforms except "__APPLE__".
    // If you want to force usage of 'poll', pass "-DHAVE_POLL=1" to gcc.
```
and [perhaps this](https://beesbuzz.biz/code/5739-The-problem-with-select-vs-poll) and maybe this as well #53.

pietrygamat commented on src/test/java/jssc/SerialNativeInterfaceTest.java:
```diff
@@ -1,16 +1,24 @@
 
 public class SerialNativeInterfaceTest {
 
+    private SerialNativeInterface testTarget;
+
+    @Before
+    public void before(){
+        testTarget = new SerialNativeInterface();
+    }
```
This should best be moved from `@Before` directly to `throwsIllegalArgumentExceptionIfPortHandleIllegal()` method - just as it's done in `reportsWriteErrorsAsIOException`. We have no need to rerun this block for each test in this class, if only one is ever going to make use of it.

hiddenalpha committed:
Apply review changes (PR 155)

Round 1: - Fix typo in exception msg.
         - Indent 'case' more.
         - use variable in place of constant.
Round 2: - Ignore test on windoof.
         - Whops, did ignore the wrong test (force-pushed to fix the
           other one).
Round 3: - Enable windoof test again as requested in PR review.
Round 4: - Move failing OSX to its own test.

Taken from 3dcb4ecd2a78a60e83d8824398288ff5ddbe6efe.
